drivers: display: Add display_clear function

rruuaanng opened the pull request:
from #12937

vanwinkeljan:
It is suggested that a new API should be introduced to clear the display screen, rather than misusing `display_set_contrast`.

cc @vanwinkeljan 

I believe this makes sense for application developers. Sometimes, when implementing screen clearing, I have to write functions that send raw messages, so I prefer to leave the implementation to the driver vendors rather than the users.

Closes: #12937

fabiobaltieri commented on include/zephyr/drivers/display.h:
```diff
@@ -281,6 +289,26 @@ static inline int display_read(const struct device *dev, const uint16_t x,
+ */
+static inline int display_clear(const struct device *dev)
+{
+	struct display_driver_api *api =
+		(struct display_driver_api *)dev->api;
```
nit: this would fit in 80 columns, unwrap?

rruuaanng replied:
In order to synchronize the code style of other functions, I have to do this (maybe I can take this opportunity to tidy up their style.
```
static inline int display_read(const struct device *dev, const uint16_t x,
			       const uint16_t y,
			       const struct display_buffer_descriptor *desc,
			       void *buf)
{
	struct display_driver_api *api =
		(struct display_driver_api *)dev->api;

	if (api->read == NULL) {
		return -ENOSYS;
	}

	return api->read(dev, x, y, desc, buf);
}
```

soburi commented:
Although it is simply called "clear," its behavior is not uniquely determined.
It would be better to make it more clear.
For example, setting memory to 0 (in this case, the colors will be inverted for MONO01 and MONO10), making the screen black...

Because of these difficulties, I think including a reference implementation with SDL in this PR would be a good idea.

soburi commented on include/zephyr/drivers/display.h:
```diff
@@ -165,6 +165,13 @@ typedef int (*display_read_api)(const struct device *dev, const uint16_t x,
+/**
+ * @typedef display_clear
+ * @brief Callback API for clearing the screen of the display
+ * See display_clear() for argument description
```
Although it is simply called "clear," its behavior is not uniquely determined.
It would be better to make it more clear.
For example, setting memory to 0 (in this case, the colors will be inverted for MONO01 and MONO10), making the screen black...
Perhaps it would be better to call the action of filling the entire screen with a specific color "clear" (I prefer this).

soburi replied:
Because of these difficulties, I think including a reference implementation with SDL in this PR would be a good idea.

rruuaanng replied:
> For example, setting memory to 0 (in this case, the colors will be inverted for MONO01 and MONO10), making the screen black...

The default clearing behavior can be defined as filling the screen with solid black(I think)—unless inversion is enabled, in which case it would be white. In other words, I believe this should be determined by the driver implementer. If they consider the "default" state of the screen to be white or any other color, that should be acceptable.

> Because of these difficulties, I think including a reference implementation with SDL in this PR would be a good idea.

I believe `display_sdl.c` is an appropriate position.

soburi replied:
When designing a HAL, it is necessary to find a "greatest common denominator" that allows uniform behavior on any hardware and is compatible with many hardware devices.
This is also the valuable benefit of using Zephyr.
Therefore, "depends on implementation" is a kind of resignation, and the design must avoid this.

IAmAnkur commented on drivers/display/display_sdl.c:
```diff
@@ -428,6 +428,44 @@ static int sdl_display_read(const struct device *dev, const uint16_t x,
+
+	switch (disp_data->current_pixel_format) {
+	case PIXEL_FORMAT_ARGB_8888:
+		size = config->width * config->height * 4U;
```
LOG_DBG of the value being set would help in Debug.

rruuaanng replied:
That to say, like this?
```
LOG_DBG("size: %zu", size);
```

IAmAnkur replied:
Yes could be at a single place, right before memset : something like this: LOG_DBG("size: %zu, bgcolor: %hhu", size, bgcolor);

rruuaanng replied:
Done. I have added it before the `memset`.

rruuaanng commented:
Change log:

- Move the migration entry to 4.2.
- Fix the incorrect fill color, black in RGB is 0x00, not 0xFF.

soburi commented:
@rruuaanng 
Could you rebase it?

rruuaanng committed:
drivers: display: Add display_clear function

Add a display_clear function to the display subsystem
to clear the target devices screen display. Introduce
a clear interface at the driver layer to implement
the display_clear function.

Signed-off-by: James Roy <rruuaanng@outlook.com>

soburi commented on drivers/display/display_sdl.c:
```diff
@@ -428,6 +428,43 @@ static int sdl_display_read(const struct device *dev, const uint16_t x,
 
+static int sdl_display_clear(const struct device *dev)
+{
+	size_t size = 0;
```
This is not blocking, but the The "reverse Christmas tree" notation is more popular.

```
	const struct sdl_display_config *config = dev->config;
	struct sdl_display_data *disp_data = dev->data;
	uint8_t bgcolor = 0x00u;
	size_t size = 0;
```

soburi commented on doc/releases/release-notes-4.2.rst:
```diff
@@ -69,6 +69,11 @@ New APIs and options
 
     * :c:func:`bt_le_get_local_features`
 
+* Display
```
@fabiobaltieri @danieldegrasse @jfischer-no 

Overall, I don't think there's any problem.
Could we hear your opinions?

fabiobaltieri commented on doc/releases/release-notes-4.2.rst:
```diff
@@ -69,6 +69,11 @@ New APIs and options
 
+* Display
+
+  * Add function ``display_clear`` to the display subsystem to clear the target devices
+    screen display, resetting it to black. (:github:`12937`)
```
just

```
:c:func:`bt_le_get_local_features`
```

the idea is that the details are already in the linked documentation, less duplication work when api gets added

fabiobaltieri commented on doc/releases/release-notes-4.2.rst:
this commit should be the last one, not the first, the function should be implemented before trying to link it in the doc (or in the same commit if you prefer I guess)

fabiobaltieri commented on drivers/display/display_sdl.c:
```diff
@@ -428,6 +428,43 @@ static int sdl_display_read(const struct device *dev, const uint16_t x,
+	switch (disp_data->current_pixel_format) {
+	case PIXEL_FORMAT_ARGB_8888:
+		size = config->width * config->height * 4U;
+		bgcolor = 0xFFu;
```
`UINT8_MAX`?

fabiobaltieri commented on drivers/display/display_sdl.c:
```diff
@@ -428,6 +428,43 @@ static int sdl_display_read(const struct device *dev, const uint16_t x,
+{
+	const struct sdl_display_config *config = dev->config;
+	struct sdl_display_data *disp_data = dev->data;
+	size_t size = 0;
```
drop the initialization from this one, that way the compiler fires a warning if there's a code path that uses it without assigning a values

rruuaanng committed:
drivers: display: Add SDL implementation example

Implement the '.clear' interface in the SDL driver
to allow the upper layers to use the display_clear
function.

Signed-off-by: James Roy <rruuaanng@outlook.com>

fabiobaltieri commented on doc/releases/release-notes-4.2.rst:
```diff
@@ -81,6 +81,11 @@ New APIs and options
 
+* Display
+
+  * Add function :c:func:`display_clear` to the display subsystem to clear the target devices
+    screen display, resetting it to black. (:github:`12937`)
```
```suggestion
  * :c:func:`display_clear`
```

rruuaanng committed:
doc: releases: Add a display release entry

Add the display_clear function in the display
subsystem to clear the target devices screen
display by resetting it to white.

Signed-off-by: James Roy <rruuaanng@outlook.com>

danieldegrasse reviewed (commented):
No issue overall with this change. I guess the idea here is that some controllers have a specific instruction to clear the display? If possible, I'd prefer to see an implementation of this in an upstream display controller driver (but if there isn't one upstream that would use it, that is ok)
